internal/controllers/bmc_reconciler.go

```diff
@@ -0,0 +1,152 @@
+	helper := newBMCReconcilerHelper(client, mcfgAPI, scheme)
+	return &bmcReconciler{
+		reconHelper: helper,
+	}
+}
+
+func (r *bmcReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kmmv1beta1.BootModuleConfig{}).
+		Named(BMCReconcilerName).
+		Owns(&apimcfgv1.MachineConfig{}).
+		Complete(
+			reconcile.AsReconciler[*kmmv1beta1.BootModuleConfig](mgr.GetClient(), r),
+		)
+}
+
+func (r *bmcReconciler) Reconcile(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Starting BootModuleConfig reconciliation")
+
+	if bmc.GetDeletionTimestamp() != nil {
+		// bootmoduleconfig is being deleted
+		err := r.reconHelper.finalizeBMC(ctx, bmc)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to finalize BootModuleConfig %s/%s: %v", bmc.Namespace, bmc.Name, err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	err := r.reconHelper.setFinalizer(ctx, bmc)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set finalize on BootModuleConfig %s/%s: %v", bmc.Namespace, bmc.Name, err)
+	}
+
+	// handling MachineConfiguration must come before handling MachineConfig in order to avoid reboots
+	err = r.reconHelper.handleMachineConfiguration(ctx, bmc)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to handle MachineConfiguration for BootModuleConfig %s/%s: %v", bmc.Namespace, bmc.Name, err)
+	}
+
+	err = r.reconHelper.handleMachineConfig(ctx, bmc)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to handle MachineConfig for BootModuleConfig %s/%s: %v", bmc.Namespace, bmc.Name, err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+//go:generate mockgen -source=bmc_reconciler.go -package=controllers -destination=mock_bmc_reconciler.go bmcReconcilerHelperAPI
+
+type bmcReconcilerHelperAPI interface {
+	finalizeBMC(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) error
+	setFinalizer(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) error
+	handleMachineConfiguration(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) error
+	handleMachineConfig(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) error
+}
+
+type bmcReconcilerHelper struct {
+	client  client.Client
+	mcfgAPI mcfg.MCFG
+	scheme  *runtime.Scheme
+}
+
+func newBMCReconcilerHelper(client client.Client, mcfgAPI mcfg.MCFG, scheme *runtime.Scheme) bmcReconcilerHelperAPI {
+	return &bmcReconcilerHelper{
+		client:  client,
+		mcfgAPI: mcfgAPI,
+		scheme:  scheme,
+	}
+}
+
+func (brh *bmcReconcilerHelper) setFinalizer(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) error {
+	if controllerutil.ContainsFinalizer(bmc, constants.BMCFinalizer) {
+		return nil
+	}
+
+	bmcCopy := bmc.DeepCopy()
+	controllerutil.AddFinalizer(bmcCopy, constants.BMCFinalizer)
+	return brh.client.Patch(ctx, bmc, client.MergeFrom(bmcCopy))
+}
+
+func (brh *bmcReconcilerHelper) finalizeBMC(ctx context.Context, bmc *kmmv1beta1.BootModuleConfig) error {
```

Comment:
Any reason to push it before we have implemented it?

Reply:
yes. Pushing it will allow Costi to start testing it. The finalization issue is something to think about, i am not sure yet if and when to use it